koop serve fails when path has spaces

`koop serve` fails when directory has spaces in name.   IE `/my project/koop`.

```
Error: Cannot find module '/my '
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:581:15)
    at Function.Module._load (internal/modules/cjs/loader.js:507:25)
    at Function.Module.runMain (internal/modules/cjs/loader.js:742:12)
    at startup (internal/bootstrap/node.js:283:19)
    at bootstrapNodeJSCore (internal/bootstrap/node.js:743:3)
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -4,6 +4,8 @@ All notable changes to this project will be documented in this file.
 This project adheres to [Semantic Versioning](http://semver.org/).
 
 ## Unreleased
+### Fixed
+* A path with spaces can crash the `serve` command (#42)
 
 ## 0.6.0 - 2019-12-09
 

--- a/package.json
+++ b/package.json
@@ -7,8 +7,7 @@
     "test": "npm run test:cli && npm run test:template",
     "test:template": "cross-env SUPPRESS_NO_CONFIG_WARNING=true mocha \"src/template/components/**/test/**/*.test.js\"",
     "test:cli": "cross-env SUPPRESS_NO_CONFIG_WARNING=true mocha \"test/**/*.test.js\"",
-    "ncu": "node scripts/ncu.js",
-    "precommit": "lint-staged && npm test"
+    "ncu": "node scripts/ncu.js"
   },
   "repository": {
     "type": "git",
@@ -61,5 +60,10 @@
       "standard --fix",
       "git add"
     ]
+  },
+  "husky": {
+    "hooks": {
+      "pre-commit": "lint-staged && npm test"
+    }
   }
 }

--- a/package.json
+++ b/package.json
@@ -27,6 +27,7 @@
   },
   "homepage": "https://github.com/koopjs/koop-cli#readme",
   "dependencies": {
+    "dargs": "^7.0.0",
     "execa": "^3.4.0",
     "fs-extra": "^8.1.0",
     "klaw-sync": "^6.0.0",

--- a/src/utils/serve/get-path-arg.js
+++ b/src/utils/serve/get-path-arg.js
@@ -3,15 +3,18 @@ const { execSync } = require('child_process')
 
 // This function wraps a path into a command line argument. This is necessary
 // because spaces in the path can break the command.
-module.exports = (path) => {
+module.exports = (path, isWatchMode) => {
   if (os.platform() === 'win32') {
     // For win, the path is resolved to DOS 8.3 format (with no space). See the
     // following link for details:
     // https://github.com/botpress/botpress/pull/338
     const shortPath = execSync(`@echo off && for %I in ("${path}") do echo %~sI`)
     return shortPath.toString('utf8').replace(/(\n|\r)+$/, '')
-  } else {
-    // For linux/mac, simply quoting the path should work.
+  } else if (isWatchMode) {
+    // For linux/mac, simply quoting the path should work (this works for nodemon)
     return `"${path}"`
+  } else {
+    // but execa doesn't like quote in its path
+    return path
   }
 }

--- a/src/utils/serve/get-path-arg.js
+++ b/src/utils/serve/get-path-arg.js
@@ -1,0 +1,17 @@
+const os = require('os')
+const { execSync } = require('child_process')
+
+// This function wraps a path into a command line argument. This is necessary
+// because spaces in the path can break the command.
+module.exports = (path) => {
+  if (os.platform() === 'win32') {
+    // For win, the path is resolved to DOS 8.3 format (with no space). See the
+    // following link for details:
+    // https://github.com/botpress/botpress/pull/338
+    const shortPath = execSync(`@echo off && for %I in ("${path}") do echo %~sI`)
+    return shortPath.toString('utf8').replace(/(\n|\r)+$/, '')
+  } else {
+    // For linux/mac, simply quoting the path should work.
+    return `"${path}"`
+  }
+}

--- a/src/utils/serve/index.js
+++ b/src/utils/serve/index.js
@@ -43,7 +43,7 @@ module.exports = async (cwd, options = {}) => {
       throw new Error('A GeoJSON file is requried to provide test data for the dev server.')
     }
 
-    const serverPath = getPathArg(path.join(dirname, './serve-plugin.js'), options.watch)
+    const serverPath = getPathArg(path.join(dirname, 'serve-plugin.js'), options.watch)
     const cwdPath = getPathArg(cwd, options.watch)
     nodeArgs._.push(serverPath)
     serveArgs.cwd = cwdPath

--- a/src/utils/serve/index.js
+++ b/src/utils/serve/index.js
@@ -3,23 +3,31 @@ const fs = require('fs-extra')
 const execa = require('execa')
 const nodemon = require('nodemon')
 const os = require('os')
+const dargs = require('dargs')
 const getPathArg = require('./get-path-arg')
 
 module.exports = async (cwd, options = {}) => {
   const koopConfig = await fs.readJson(path.join(cwd, 'koop.json'))
   // for test mocking
   const dirname = options.dirname || __dirname
-  let command = ''
+  const nodeArgs = {
+    _: []
+  }
+  const serveArgs = {}
+
+  if (options.debug) {
+    nodeArgs.inspect = true
+  }
 
   if (options.path) {
     // run the test server file if provided
-    const filePath = getPathArg(path.join(cwd, options.path))
-    command = filePath
+    const filePath = getPathArg(options.path, options.watch)
+    nodeArgs._.push(filePath)
   } else if (koopConfig.type === 'app') {
     // if it is an app, run it directly
     const packageInfo = await fs.readJson(path.join(cwd, 'package.json'))
-    const appPath = getPathArg(path.join(cwd, packageInfo.main))
-    command = appPath
+    const appPath = getPathArg(packageInfo.main, options.watch)
+    nodeArgs._.push(appPath)
   } else {
     // otherwise, this is a plugin and we should run a Koop server for it
 
@@ -35,43 +43,43 @@ module.exports = async (cwd, options = {}) => {
       throw new Error('A GeoJSON file is requried to provide test data for the dev server.')
     }
 
-    const serverPath = getPathArg(path.join(dirname, './serve-plugin'))
-    const cwdPath = getPathArg(cwd)
-    command = `${serverPath} --cwd=${cwdPath}`
+    const serverPath = getPathArg(path.join(dirname, './serve-plugin.js'), options.watch)
+    const cwdPath = getPathArg(cwd, options.watch)
+    nodeArgs._.push(serverPath)
+    serveArgs.cwd = cwdPath
 
     if (options.data) {
-      const dataPath = getPathArg(options.data)
-      command += ` --data-path=${dataPath}`
+      const dataPath = getPathArg(options.data, options.watch)
+      serveArgs.dataPath = dataPath
     }
 
     if (options.port) {
-      command += ` --port=${options.port}`
+      serveArgs.port = options.port
     }
   }
 
-  if (options.debug) {
-    command = `--inspect ${command}`
-  }
-
-  command = `node ${command}`
+  const args = [
+    ...dargs(nodeArgs),
+    ...dargs(serveArgs)
+  ]
 
   if (options.watch) {
-    startsWithNodemon(cwd, command)
+    startsWithNodemon(cwd, args)
   } else {
-    startsWithNode(cwd, command)
+    startsWithNode(cwd, args)
   }
 }
 
-function startsWithNode (cwd, command) {
-  execa.command(command, {
+function startsWithNode (cwd, args) {
+  execa('node', args, {
     cwd,
     stdio: 'inherit'
   })
 }
 
-function startsWithNodemon (cwd, command) {
+function startsWithNodemon (cwd, args) {
   nodemon({
-    exec: command,
+    exec: `node ${args.join(' ')}`,
     watch: cwd,
     ignore: ['node_modules', '.git']
   })

--- a/src/utils/serve/index.js
+++ b/src/utils/serve/index.js
@@ -3,6 +3,7 @@ const fs = require('fs-extra')
 const execa = require('execa')
 const nodemon = require('nodemon')
 const os = require('os')
+const getPathArg = require('./get-path-arg')
 
 module.exports = async (cwd, options = {}) => {
   const koopConfig = await fs.readJson(path.join(cwd, 'koop.json'))
@@ -12,11 +13,13 @@ module.exports = async (cwd, options = {}) => {
 
   if (options.path) {
     // run the test server file if provided
-    command = path.join(cwd, options.path)
+    const filePath = getPathArg(path.join(cwd, options.path))
+    command = filePath
   } else if (koopConfig.type === 'app') {
     // if it is an app, run it directly
     const packageInfo = await fs.readJson(path.join(cwd, 'package.json'))
-    command = path.join(cwd, packageInfo.main)
+    const appPath = getPathArg(path.join(cwd, packageInfo.main))
+    command = appPath
   } else {
     // otherwise, this is a plugin and we should run a Koop server for it
 
@@ -32,10 +35,13 @@ module.exports = async (cwd, options = {}) => {
       throw new Error('A GeoJSON file is requried to provide test data for the dev server.')
     }
 
-    command = `${path.join(dirname, './serve-plugin')} --cwd=${cwd}`
+    const serverPath = getPathArg(path.join(dirname, './serve-plugin'))
+    const cwdPath = getPathArg(cwd)
+    command = `${serverPath} --cwd=${cwdPath}`
 
     if (options.data) {
-      command += ` --data-path=${options.data}`
+      const dataPath = getPathArg(options.data)
+      command += ` --data-path=${dataPath}`
     }
 
     if (options.port) {

--- a/test/utils/serve/get-path-arg.test.js
+++ b/test/utils/serve/get-path-arg.test.js
@@ -9,12 +9,20 @@ const expect = chai.expect
 const getPathArg = require('../../../src/utils/serve/get-path-arg')
 
 describe('utils/serve/get-path-arg', () => {
-  it('should quote the path in Linux/Max', function () {
+  it('should not quote the path in Linux/Max without watch mode', function () {
     if (os.platform() === 'win32') {
       this.skip()
     }
 
-    expect(getPathArg('some path')).to.equal('"some path"')
+    expect(getPathArg('some path')).to.equal('some path')
+  })
+
+  it('should quote the path in Linux/Max in watch mode', function () {
+    if (os.platform() === 'win32') {
+      this.skip()
+    }
+
+    expect(getPathArg('some path', true)).to.equal('"some path"')
   })
 
   it('should return a DOS short path with no space in Windows', async function () {

--- a/test/utils/serve/get-path-arg.test.js
+++ b/test/utils/serve/get-path-arg.test.js
@@ -22,7 +22,7 @@ describe('utils/serve/get-path-arg', () => {
       this.skip()
     }
 
-    const originPath = path.join(os.tmpdir, 'space space', 'secret.txt')
+    const originPath = path.join(os.tmpdir(), 'space space', 'secret.txt')
     const secretText = 'my secret'
     await fs.outputFile(originPath, secretText)
 

--- a/test/utils/serve/get-path-arg.test.js
+++ b/test/utils/serve/get-path-arg.test.js
@@ -1,0 +1,35 @@
+/* eslint-env mocha */
+
+const os = require('os')
+const path = require('path')
+const fs = require('fs-extra')
+const chai = require('chai')
+const expect = chai.expect
+
+const getPathArg = require('../../../src/utils/serve/get-path-arg')
+
+describe('utils/serve/get-path-arg', () => {
+  it('should quote the path in Linux/Max', function () {
+    if (os.platform === 'win32') {
+      this.skip()
+    }
+
+    expect(getPathArg('some path')).to.equal('"some path"')
+  })
+
+  it('should return a DOS short path with no space in Windows', async function () {
+    if (os.platform !== 'win32') {
+      this.skip()
+    }
+
+    const originPath = path.join(os.tmpdir, 'space space', 'secret.txt')
+    const secretText = 'my secret'
+    await fs.outputFile(originPath, secretText)
+
+    const newPath = getPathArg(originPath)
+    expect(newPath).to.not.include(' ')
+
+    // verify the content again
+    expect(await fs.readFile(newPath, 'utf-8')).to.equal(secretText)
+  })
+})

--- a/test/utils/serve/get-path-arg.test.js
+++ b/test/utils/serve/get-path-arg.test.js
@@ -10,7 +10,7 @@ const getPathArg = require('../../../src/utils/serve/get-path-arg')
 
 describe('utils/serve/get-path-arg', () => {
   it('should quote the path in Linux/Max', function () {
-    if (os.platform === 'win32') {
+    if (os.platform() === 'win32') {
       this.skip()
     }
 
@@ -18,7 +18,7 @@ describe('utils/serve/get-path-arg', () => {
   })
 
   it('should return a DOS short path with no space in Windows', async function () {
-    if (os.platform !== 'win32') {
+    if (os.platform() !== 'win32') {
       this.skip()
     }
 

--- a/test/utils/serve/index.test.js
+++ b/test/utils/serve/index.test.js
@@ -11,7 +11,6 @@
 
 const proxyquire = require('proxyquire')
 const chai = require('chai')
-const path = require('path')
 
 const expect = chai.expect
 const moduleName = '../../../src/utils/serve'
@@ -19,11 +18,10 @@ const moduleName = '../../../src/utils/serve'
 describe('utils/serve', () => {
   it('should run the given server file', async () => {
     const serve = proxyquire(moduleName, {
-      execa: {
-        command (command) {
-          const expectedPath = path.join('/', 'app.js')
-          expect(command).to.equal(`node ${expectedPath}`)
-        }
+      execa: (command, args, options) => {
+        expect(command).to.equal('node')
+        expect(args).to.deep.equal(['app.js'])
+        expect(options.cwd).to.equal('/')
       },
       'fs-extra': {
         async readJson () {
@@ -38,11 +36,10 @@ describe('utils/serve', () => {
 
   it('should run an app', async () => {
     const serve = proxyquire(moduleName, {
-      execa: {
-        command (command) {
-          const expectedPath = path.join('/', 'src', 'index.js')
-          expect(command).to.equal(`node ${expectedPath}`)
-        }
+      execa: (command, args, options) => {
+        expect(command).to.equal('node')
+        expect(args).to.deep.equal(['src/index.js'])
+        expect(options.cwd).to.equal('/')
       },
       'fs-extra': {
         async readJson (path) {
@@ -61,11 +58,10 @@ describe('utils/serve', () => {
 
   it('should run a provider', async () => {
     const serve = proxyquire(moduleName, {
-      execa: {
-        command (command) {
-          const expectedPath = path.join('/', 'serve-plugin')
-          expect(command).to.equal(`node ${expectedPath} --cwd=/`)
-        }
+      execa: (command, args, options) => {
+        expect(command).to.equal('node')
+        expect(args).to.deep.equal(['/serve-plugin.js', '--cwd=/'])
+        expect(options.cwd).to.equal('/')
       },
       'fs-extra': {
         async readJson (path) {
@@ -84,11 +80,10 @@ describe('utils/serve', () => {
 
   it('should run an output', async () => {
     const serve = proxyquire(moduleName, {
-      execa: {
-        command (command) {
-          const expectedPath = path.join('/', 'serve-plugin')
-          expect(command).to.equal(`node ${expectedPath} --cwd=/ --data-path=test.geojson`)
-        }
+      execa: (command, args, options) => {
+        expect(command).to.equal('node')
+        expect(args).to.deep.equal(['/serve-plugin.js', '--cwd=/', '--data-path=test.geojson'])
+        expect(options.cwd).to.equal('/')
       },
       'fs-extra': {
         async readJson (path) {
@@ -110,11 +105,10 @@ describe('utils/serve', () => {
 
   it('should run an auth', async () => {
     const serve = proxyquire(moduleName, {
-      execa: {
-        command (command) {
-          const expectedPath = path.join('/', 'serve-plugin')
-          expect(command).to.equal(`node ${expectedPath} --cwd=/ --data-path=test.geojson`)
-        }
+      execa: (command, args, options) => {
+        expect(command).to.equal('node')
+        expect(args).to.deep.equal(['/serve-plugin.js', '--cwd=/', '--data-path=test.geojson'])
+        expect(options.cwd).to.equal('/')
       },
       'fs-extra': {
         async readJson (path) {
@@ -156,11 +150,10 @@ describe('utils/serve', () => {
 
   it('could run in debug mode', async () => {
     const serve = proxyquire(moduleName, {
-      execa: {
-        command (command) {
-          const expectedPath = path.join('/', 'app.js')
-          expect(command).to.equal(`node --inspect ${expectedPath}`)
-        }
+      execa: (command, args, options) => {
+        expect(command).to.equal('node')
+        expect(args).to.deep.equal(['--inspect', 'app.js'])
+        expect(options.cwd).to.equal('/')
       },
       'fs-extra': {
         async readJson () {
@@ -176,8 +169,7 @@ describe('utils/serve', () => {
   it('could run in watch mode', (done) => {
     const serve = proxyquire(moduleName, {
       nodemon: (config) => {
-        const expectedPath = path.join('/', 'app.js')
-        expect(config.exec).to.equal(`node ${expectedPath}`)
+        expect(config.exec).to.equal('node app.js')
         expect(config.watch).to.equal('/')
 
         return {

--- a/test/utils/serve/index.test.js
+++ b/test/utils/serve/index.test.js
@@ -29,7 +29,8 @@ describe('utils/serve', () => {
         async readJson () {
           return {}
         }
-      }
+      },
+      './get-path-arg': (path) => path
     })
 
     return serve('/', { path: 'app.js' })
@@ -51,7 +52,8 @@ describe('utils/serve', () => {
             return { main: 'src/index.js' }
           }
         }
-      }
+      },
+      './get-path-arg': (path) => path
     })
 
     return serve('/')
@@ -73,7 +75,8 @@ describe('utils/serve', () => {
             return { main: 'src/index.js' }
           }
         }
-      }
+      },
+      './get-path-arg': (path) => path
     })
 
     return serve('/', { dirname: '/' })
@@ -98,7 +101,8 @@ describe('utils/serve', () => {
         async pathExists () {
           return true
         }
-      }
+      },
+      './get-path-arg': (path) => path
     })
 
     return serve('/', { data: 'test.geojson', dirname: '/' })
@@ -123,7 +127,8 @@ describe('utils/serve', () => {
         async pathExists () {
           return true
         }
-      }
+      },
+      './get-path-arg': (path) => path
     })
 
     return serve('/', { data: 'test.geojson', dirname: '/' })
@@ -138,7 +143,8 @@ describe('utils/serve', () => {
         async pathExists () {
           return false
         }
-      }
+      },
+      './get-path-arg': (path) => path
     })
 
     try {
@@ -160,7 +166,8 @@ describe('utils/serve', () => {
         async readJson () {
           return {}
         }
-      }
+      },
+      './get-path-arg': (path) => path
     })
 
     return serve('/', { path: 'app.js', debug: true })
@@ -183,7 +190,8 @@ describe('utils/serve', () => {
         async readJson () {
           return {}
         }
-      }
+      },
+      './get-path-arg': (path) => path
     })
 
     serve('/', { path: 'app.js', watch: true })

--- a/test/utils/serve/index.test.js
+++ b/test/utils/serve/index.test.js
@@ -11,7 +11,7 @@
 
 const proxyquire = require('proxyquire')
 const chai = require('chai')
-
+const path = require('path')
 const expect = chai.expect
 const moduleName = '../../../src/utils/serve'
 
@@ -60,7 +60,7 @@ describe('utils/serve', () => {
     const serve = proxyquire(moduleName, {
       execa: (command, args, options) => {
         expect(command).to.equal('node')
-        expect(args).to.deep.equal(['/serve-plugin.js', '--cwd=/'])
+        expect(args).to.deep.equal([path.join('test', 'serve-plugin.js'), '--cwd=/'])
         expect(options.cwd).to.equal('/')
       },
       'fs-extra': {
@@ -75,14 +75,14 @@ describe('utils/serve', () => {
       './get-path-arg': (path) => path
     })
 
-    return serve('/', { dirname: '/' })
+    return serve('/', { dirname: 'test' })
   })
 
   it('should run an output', async () => {
     const serve = proxyquire(moduleName, {
       execa: (command, args, options) => {
         expect(command).to.equal('node')
-        expect(args).to.deep.equal(['/serve-plugin.js', '--cwd=/', '--data-path=test.geojson'])
+        expect(args).to.deep.equal([path.join('test', 'serve-plugin.js'), '--cwd=/', '--data-path=test.geojson'])
         expect(options.cwd).to.equal('/')
       },
       'fs-extra': {
@@ -100,14 +100,14 @@ describe('utils/serve', () => {
       './get-path-arg': (path) => path
     })
 
-    return serve('/', { data: 'test.geojson', dirname: '/' })
+    return serve('/', { data: 'test.geojson', dirname: 'test' })
   })
 
   it('should run an auth', async () => {
     const serve = proxyquire(moduleName, {
       execa: (command, args, options) => {
         expect(command).to.equal('node')
-        expect(args).to.deep.equal(['/serve-plugin.js', '--cwd=/', '--data-path=test.geojson'])
+        expect(args).to.deep.equal([path.join('test', 'serve-plugin.js'), '--cwd=/', '--data-path=test.geojson'])
         expect(options.cwd).to.equal('/')
       },
       'fs-extra': {
@@ -125,7 +125,7 @@ describe('utils/serve', () => {
       './get-path-arg': (path) => path
     })
 
-    return serve('/', { data: 'test.geojson', dirname: '/' })
+    return serve('/', { data: 'test.geojson', dirname: 'test' })
   })
 
   it('should throw an error for missing test data file', async () => {

